halo2-base/src/gates/tests/prop_test.rs

```diff
@@ -0,0 +1,333 @@
+            idx_value = i as usize;
+            break;
+        }
+    }
+    if idx_value < inputs.0.len() {
+        indicator[idx_value] = F::one();
+    }
+    // take cross product of indicator and inputs.0
+    inputs.0.iter().zip(indicator.iter()).fold(F::zero(),|acc, (a, b)| acc + (*a.value() * *b))
+}
+
+fn select_from_idx_ground_truth<F: ScalarField>(inputs: &(Vec<QuantumCell<F>>, QuantumCell<F>)) -> F {
+    let idx = inputs.1.value();
+    // Since F does not implement From<u64>, we have to iterate and find the matching index
+    for i in 0..inputs.0.len() as u64 {
+        if F::from(i) == *idx {
+            return *inputs.0[i as usize].value();
+        }
+    }
+    F::zero()
+}
+
+fn is_zero_ground_truth<F: ScalarField>(x: F) -> F {
+    if x.is_zero().into() { F::one() } else { F::zero() }
+}
+
+fn is_equal_ground_truth<F: ScalarField>(inputs: &[QuantumCell<F>]) -> F {
+    if inputs[0].value() == inputs[1].value() {
+        F::one()
+    } else {
+        F::zero()
+    }
+}
+
+fn lagrange_eval_ground_truth<F: ScalarField>(inputs: &[F]) -> (F, F) {
+    let x1 = inputs[0];
+    let x2 = inputs[1];
+    let y1 = inputs[2];
+    let quotient = (y1 - x1).invert().unwrap() * (x2 - x1);
+    let y2 = quotient * (x2 - x1) + x1;
+    (quotient, y2)
+}
+
+fn get_field_element_ground_truth<F: ScalarField>(n: u64) -> F {
+    F::from(n)
+}
+
+proptest! {
+
+    // Positive tests
+    #[test]
+    fn prop_test_add(input in vec(rand_witness(), 2)) {
+        let ground_truth = add_ground_truth(input.as_slice());
+        let result = flex_gate_test::test_add(input.as_slice());
+        prop_assert_eq!(result, ground_truth);
+    }
+
+    #[test]
+    fn prop_test_sub(input in vec(rand_witness(), 2)) {
+        let ground_truth = sub_ground_truth(input.as_slice());
+        let result = flex_gate_test::test_sub(input.as_slice());
+        prop_assert_eq!(result, ground_truth);
+    }
+
+    #[test]
+    fn prop_test_neg(input in rand_witness()) {
+        let ground_truth = neg_ground_truth(input);
+        let result = flex_gate_test::test_neg(input);
+        prop_assert_eq!(result, ground_truth);
+    }
+
+    #[test]
+    fn prop_test_mul(inputs in vec(rand_witness(), 2)) {
+        let ground_truth = mul_ground_truth(inputs.as_slice());
+        let result = flex_gate_test::test_mul(inputs.as_slice());
+        prop_assert_eq!(result, ground_truth);
+    }
+
+    #[test]
+    fn prop_test_mul_add(inputs in vec(rand_witness(), 3)) {
+        let ground_truth = mul_add_ground_truth(inputs.as_slice());
+        let result = flex_gate_test::test_mul_add(inputs.as_slice());
+        prop_assert_eq!(result, ground_truth);
+    }
+
+    #[test]
+    fn prop_test_mul_not(inputs in vec(rand_witness(), 2)) {
+        let ground_truth = mul_not_ground_truth(inputs.as_slice());
+        let result = flex_gate_test::test_mul_not(inputs.as_slice());
+        prop_assert_eq!(result, ground_truth);
+    }
+
+    #[test]
```

Comment:
this test currently fails because input is random `Fr`, so when it's not a bit it fails